src/batchnorm.jl

```diff
@@ -0,0 +1,485 @@
+"""
+
+`BNMoments` is a high-level data structure used to store running mean and running variance
+of batch normalization.
+
+# Fields
+ `momentum::AbstractFloat`: A real number between 0 and 1 to be used as the scale of
+  last mean and variance. The existing running mean or variance is multiplied with 
+  (1-momentum).
+ 
+ `mean`: The running mean.
+
+ `var`: The running variance.
+ 
+ `meaninit`: The function used for initialize the running mean. Should either be `nothing` or
+of the form `(eltype, dims...) -> data`. `zeros` is a good option.
+
+ `varinit`: The function used for initialize the running variance. Should either be `nothing` or
+`(eltype, dims...) -> data`. `ones` is a good option.
+
+# Constructors
+ `BNMoments(;momentum=0.1, meaninit=zeros, varinit=ones)` sets `mean` and `var` to nothing.
+When this constructor is used, `batchnorm` functions will initialize `mean` and `var` during
+the first training iteration to have relevant dimensions and match the data type of the input.
+
+ `BNMoments(momentum, mean, var)` can be used directly load moments from data. `meaninit` and
+`varinit` are made nothing and they won't be called by `batchnorm`. Its user's responsibility
+ to initialize mean and var with right dimensions and types.
+
+"""
+type BNMoments
+    momentum::AbstractFloat
+    mean
+    var
+    meaninit
+    varinit
+end
+
+BNMoments(;momentum=0.1, meaninit=zeros, varinit=ones, o...) =
+    BNMoments(momentum, nothing, nothing, meaninit, varinit)
+
+BNMoments(momentum, mean, var) = BNMoments(momentum, mean, var, nothing, nothing)
+
+
+#TODO: improve the documentation. Explain what batchnorm is.
+"""
+
+`batchnorm(x[, moments, param]; kwargs...)` performs batch normalization to `x`
+with optional scaling factor and bias stored in `w`. Operation performed is spatial 
+batch normalization if x is 4d or 5d. `moments` is a `BNMoments` object that stores 
+running mean and variance to be used in testing. It is optional in training mode, 
+but mendatory in test mode. `param` stores the optional affine parameters gamma and beta.
+`bnparam` can be used to initialize `param`.
+
+# Keywords
+
+ `eps=1e-5`: The epsilon parameter added to the variance to avoid division by 0
+ 
+ `training=nothing`: When `training` is true, the mean and variance of `x` are used and `moments`
+ argument is modified if it is provided. When `training` is false, mean and variance stored in 
+ the `moments` argument are used. When training is kept `nothing`, the mode is determined
+ based on inputs' being recorded.
+
+"""
+function batchnorm(x, moments::Union{BNMoments, Void}=nothing, param=nothing;
+                   training=nothing, o...)
+    xnd = ndims(x)
```

Comment:
Do we need o... for anything in keyword args?

Reply:
There were some options like `cache_verbose` that I used for debugging. There are also `cudnnhandle`, `alpha`, 'beta` etc. for the GPU version, which I did not documented. I added o... since `batchnorm` is a wrapper over primitives, so o... may make it easier to integrate more primitive keyword arguments to the public API. 

I don't think removing o... would break anything.